public/percent-complete/percent-complete.js

```diff
@@ -0,0 +1,265 @@
+import { JiraIssue } from "../shared/issue-data/issue-data.js";
+import { estimateExtraPoints } from "../confidence.js";
+import { millisecondsToDay, parseDate8601String } from "../date-helpers.js";
+
+/**
+ * @param { JiraIssue[] } issues
+ * @param { PercentCompleteOptions } options
+ */
+export function percentComplete(issues, options) {
+  /** @type {PercentCompleteOptions} */
+  const completeOptions = {
+    ...options,
+    toWholeDay: options.toWholeDay ?? Math.round,
+  };
+
+  const map = buildMap(issues, completeOptions);
+  setDurations(map, completeOptions);
+
+  console.log("percentComplete: map=", map);
+}
+
+/**
+ * @export
+ * @param {JiraIssue[]} issues
+ * @param {PercentCompleteOptions} options
+ * @returns {Record<string, CompletionIssue>>}
+ */
+export function buildMap(issues, options) {
+  return issues.reduce((acc, issue) => {
+    const key = options.getIssueKey(issue);
+    const parentKey = options.getParentKey(issue);
+
+    acc[key] = {
+      childKeys: acc[key]?.childKeys ?? [],
```

Comment:
If you sorted by issue hierarchy, you could avoid this "temp childKeys" stuff.  You could make sure you process all "parents" before any children.